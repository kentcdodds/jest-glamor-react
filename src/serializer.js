const {replaceSelectors} = require('./replace-selectors')
const {
  getNodes,
  markNodes,
  getSelectors,
  getStylesAndAllSelectors,
} = require('./utils')

function createSerializer(styleSheet, classNameReplacer) {
  function test(val) {
    return (
      val &&
      !val.serializedWithJestGlamorReact &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    )
  }

  function print(val, printer) {
    const nodes = getNodes(val)
    markNodes(nodes)
    const selectors = getSelectors(nodes)
    const {styles, allSelectors} = getStylesAndAllSelectors(
      selectors,
      styleSheet,
    )
    const printedVal = printer(val)
    if (styles) {
      return replaceSelectors(
        allSelectors,
        styles,
        printedVal,
        classNameReplacer,
      )
    } else {
      return printedVal
    }
  }

  return {test, print}
}

function isBeingSerialized(node) {
  let currentNode = node

  while (currentNode) {
    if (currentNode.serializedWithJestGlamorReact) {
      return true
    }
    currentNode = currentNode.parentNode
  }
  return false
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getGlamorStyleSheet = () => require('glamor').styleSheet
const glamorSerializer = createSerializer(getGlamorStyleSheet)
createSerializer.test = glamorSerializer.test
createSerializer.print = glamorSerializer.print

module.exports = createSerializer
