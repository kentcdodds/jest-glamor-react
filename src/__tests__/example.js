import React from 'react'
import ReactDOM from 'react-dom'
import * as glamor from 'glamor'
import serializer from '../serializer'
import {toHaveStyleRule} from '../matchers'

expect.addSnapshotSerializer(serializer)
expect.extend({toHaveStyleRule})

function Wrapper(props) {
  const className = glamor.css({
    padding: '4em',
    background: 'papayawhip',
  })
  return <section className={`${className}`} {...props} />
}

function Title(props) {
  const className = glamor.css({
    fontSize: '1.5em',
    textAlign: 'center',
    color: 'palevioletred',
  })
  return <h1 className={`${className}`} {...props} />
}

function render(ui) {
  const div = document.createElement('div')
  ReactDOM.render(ui, div)
  return div.children[0]
}

test('react-dom snapshot', () => {
  const node = render(
    <Wrapper>
      <Title>Hello World, this is my first glamor styled component!</Title>
    </Wrapper>,
  )
  expect(node).toMatchSnapshot()
})

test('react-dom toHaveStyleRule', () => {
  const node = render(
    <Wrapper>
      <Title>Hello World, this is my first glamor styled component!</Title>
    </Wrapper>,
  )
  expect(node).toHaveStyleRule('background', 'papayawhip')
})
