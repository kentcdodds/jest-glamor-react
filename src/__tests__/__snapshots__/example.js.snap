// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-dom snapshot 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

.css-1,
[data-css-1] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  class="css-0"
>
  <h1
    class="css-1"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;
